Documentation/design-docs/COM-activation.md

```diff
@@ -0,0 +1,132 @@
+
+1) Determine additional registration information needed for activation.
+    * **The location and mechanics of accessing this registration information is still TBD**
+    * `Assembly` (**required**) - the [Fully-Qualified Name](https://docs.microsoft.com/en-us/dotnet/framework/reflection-and-codedom/specifying-fully-qualified-type-names) of the assembly.
+    * `Class` (**required**) - the full type name (e.g. `MyCompany.MyProduct.MyClass`).
+    * `Codebase` (**required**) - an absolute path to the assembly to load. If a [`runtimeconfig.json`](https://github.com/dotnet/cli/blob/master/Documentation/specs/runtime-configuration-file.md) file exists adjacent to the assembly that file will be preferred to describe CLR configuration details.
+    * `RuntimeVersion` (optional) - CLR version requirement.
+1) Using the existing `hostfxr` library, attempt to discover the desired CLR.
+    * If a CLR is present, the requested `RuntimeVersion` will be validated against that CLR. If version satisfiability fails, activation will fail.
+    * If a CLR is **not** present, an attempt will be made to create a satisfying CLR instance. Failure to create an instance will result in activation failure.
+1) A request to the CLR will be made via a new method for class activation within a COM environment.
+    * The ability to load the assembly and create a class instance will require an additional function be exposed that can be called from `hostfxr`.
+    * Example of a possible API in `System.Private.CoreLib` on the `Activator` class:
+        ``` csharp
+        public static class Activator
+        {
+            ...
+            #if WINDOWS
+            public static ObjectHandle CreateInstanceForCom(string assemblyName, string typeName);
+            #endif
+            ...
+        }
+        ```
+        Note this API would not be exposed outside of `System.Private.CoreLib`.
+    * The loading of the assembly will take place in a new [`AssemblyLoadContext`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.loader.assemblyloadcontext?view=netcore-2.1) for dependency isolation.
+1) The class instance will be returned to the caller of `DllGetClassObject()`.
+
+The `DllCanUnloadNow()` function will always return `S_FALSE` indicating the shim is never able to be unloaded. This matches .NET Framework semantics and can be adjusted in the future if needed.
+
+#### Class Registration
+
+Registration will depend on the application's [deployment scenario](https://docs.microsoft.com/en-us/dotnet/core/deploying/).
+
+* [Framework-dependent deployments (FDD)](https://docs.microsoft.com/en-us/dotnet/core/deploying/#framework-dependent-deployments-fdd) - This is most similar to the .NET Framework scenario and registration will be done using the shared framework. 
+* [Self-contained deployments (SCD)](https://docs.microsoft.com/en-us/dotnet/core/deploying/#self-contained-deployments-scd) - Registration will be done using the contained framework.
+
+##### Registry
```

Comment:
Another interesting problem: SxS. There are couples ways that COM activation will accidently introduce SxS
Scenario: .NET Core SxS .NET Framework
CoCreateInstance(CLSID_A, ...)   // CLSID_A exists in both  .Net FrameWork but not exists in .net core
CoCreateInstance(CLSID_B, ...)  //   CLSID_B exists in .Net FrameWork but not exists in .net core
As a result, .NET Framework runtime(CLR.dll) and .NET Core runtime(coreclr.dll) will both launch at same process.

Scenario: .NET Core SxS .NET Core
CoCreateInstance(CLSID_C, ...)   // CLSID_C exists in reg-free 
CoCreateInstance(CLSID_D, ...)  //   CLSID_B exists in register for .NET Core

Reply:
RegFree context is based on application manifest so I am unclear how we could mix the RegFree vs registry. If the application has a entry in the manifest that is what should be used. I guess a fallback would be to hit the registry, but I am unclear if that should be supported.

Reply:
>Scenario: .NET Core SxS .NET Framework

That is exactly why that is present in the activation matrix above.

Reply:
Will mention the issue with conflicting `GUID`s between .NET Framework and .NET Core.